### DISCUSSION TITLE
Feature Request: Import & Export Settings Commandline Options

### DISCUSSION THREAD

**Discussion by alex7654** [valinet/ExplorerPatcher#2696] [Ideas]
ExplorerPatcher's settings can easily be imported & exported via the Properties/About section. But I would really appreciate to have command line options for importing and exporting settings. This would enable e.g. importing settings via scripts or shortcuts and allow for automation.
The only command line options I found so far are "/uninstall" and "/extract". Would be great to also have "/import" and "/export" options.

**Comment by AlexanderPirvu** [comment]
Would be nice if this can also be done in such a way that it can be deployed with MSI or MSIX packages, so then it can allow for deployment via Intune/SCCM with the custom settings.

**Reply by pyrates999** [reply]
Please file this as a separate discussion.

**Comment by pyrates999** [comment]
The settings are stored per user.  The installer runs with elevated privileges.  So how would it know which user the settings you're trying to import to are for?

It can't be part of the EP properties application since this is how it is ran:

`C:\Windows\System32\rundll32.exe "C:\Program Files\ExplorerPatcher\ExplorerPatcher.amd64.dll",ZZGUI`

So an exe that can be called from the command line after installing EP would be best for this.

**Reply by alex7654** [reply]
I naively considered it a quite easy task, as there already is an import/export function built in. You made me realize that it is somewhat more tricky as I initially thought.

A separate exe for importing/exporting settings would be a perfect solution to be used by scripts, of course.

**Reply by plofhaan** [reply]
I suppose that the simplest way would be to implement the commandline switch in such a way that it can optionally accept multiple Usernames and/or UserSIDs that either can be specified directly on the command line after the switch itself (separated by space delimiters) or can be read from a textfile that can be specified after an added switch, like, e.g., _/userlist_ or _/ul_ maybe.

Also maybe, redesign ep_setup.exe so it runs with standard user privileges first, reads its argument list, reads the %username% environment variable so it can default to that one if no user was specified in the command, and finally self-elevates whilst passing the argument list on to itself so that it can continue to complete its task.

Even better would be to also provide an option to make it bypass the UAC prompt, i.e. by adding the ability to let it add an elevation task in Task Scheduler for this purpose. Ideally, this would also let it bypass the two elevation messages that are explained in the Wiki: https://github.com/valinet/ExplorerPatcher/wiki/Settings-management

Another nice option to add would be a _/quiet_ or _/q_ switch that causes GUI popups to be suppressed like the one that appears when you use the _/uninstall_ switch.

**Reply by pyrates999** [reply]
No, that is more complicated.  Much easier to have a separate exe that would be part of a script that is ran for each username.

And if an application can bypass UAC, so can malware.

**Reply by plofhaan** [reply]
I don't think so, no, as the exporter/importer separate utility .exe would still require to be aware of whether another instance of ep_setup.exe is running, so as to avoid possible conflicts with it, when it too can trigger a manipulatation of the settings during runtime. Therefore, from a technical design point of view, it makes better sense to centralize both this awareness function and the parts of the code that must rely on it, in the actual setup tool itself.
> And if an application can bypass UAC, so can malware.

For this reason, I would suggest to move this option to the Advanced category of the Properties window, and also add it as a commandline option so that it can be set up in an unattended fashion. People who already know how it is possible to use Task Scheduler to bypass UAC prompts should also already understand the risks, anyway in the first place, and, you still get a UAC prompt when you run ep_setup.exe, also anyway in the first place so...

**Reply by pyrates999** [reply]
Still no:

1. it is trivial to check if a process named ep_setup.exe exists.
2. the exe used for importing would be just another file that is copied to the install directory that EP is installed to: C:\Program Files\ExplorerPatcher
3. even if you had renamed ep_setup.exe to be another exe file, the install can create a registry key to indicate the status of EP whether it is currently installing, or the install had finished.  Even then, the install process is very quick and this would be an edge case.
4. the only way to bypass UAC that you are referring to is to create a scheduled task that saves your password.  Software alone is not allowed to do this because windows would require you to input your password.  So this would not even work what you are suggesting.
5. and I just realized, you need to be running as the user in order to import settings because it is stored in the registry under the username's registry that EP is running as.  It doesn't store it in HKEY_LOCAL_MACHINE.  It is stored in the username's registry and is named HKEY_CURRENT_USER.

**Reply by plofhaan** [reply]
Letting ep_setup.exe check whether another instance of ep_setup.exe is running is trivial also.
There is no need to create additional registry keys to implement this feature. Doing so would only complicate the code for no good reason, and registry keys still need to be cleaned up e.g. after a crash occurs.
Whether it is an edge case is irrelevant, when there simply is no need to add more code to make it automatically recover after some unexpected crash.
As for requiring to input your password, no, I am always logged in as an administrator so I never have to use my password. Again, people who do this should already understand the risks.
Finally, running ep_setup.exe as a different user should be trivial to those who already understand how to access and manipulate protected registry keys e.g. with Powershell.

**Reply by pyrates999** [reply]
You can't install multiple versions of EP in the same system.

Even though you are logged in already, when you create a scheduled task with elevated privileges, it will ask you for your password.

The source code is available.  If you want this feature the way you say it should be implemented, then have at it, and submit the pull request.

**Reply by plofhaan** [reply]
I think I have done it enough times to know it for a fact that programatically adding a new task in Task Scheduler to elevate some program that requires to be run elevated does NOT ask me for my password. Rather, it only causes a UAC prompt to be shown, as I am already logged in to an Administrator account that has all the permissions that are necessary for this.

Sorry, but I don't have enough time or courage to dig through all the code. I can see that the code to export/import the settings is located in GUI.c, but I don't personally feel the need to experiment with it. I am only here to discuss a few suggestions that shouldn't be too hard to implement by someone who already is familiar with the code, which I personally am not.

**Reply by pyrates999** [reply]
If you're not willing to put in the work, then I guess it's not needed that badly by you.
